euphonic/cli/powder_map.py

```diff
@@ -0,0 +1,155 @@
+
+    sections['q'].add_argument('--q-min', type=float, default=0., dest='q_min',
+                               help="Minimum |q| in 1/LENGTH_UNIT")
+    sections['q'].add_argument('--q-max', type=float, default=3., dest='q_max',
+                               help="Maximum |q| in 1/LENGTH_UNIT")
+    return parser
+
+
+def main(params: List[str] = None):
+    args = get_args(get_parser(), params)
+    calc_modes_kwargs = _calc_modes_kwargs(args)
+
+    # Make sure we get an error if accessing NPTS inappropriately
+    if args.npts_density is not None:
+        args.npts = None
+
+    fc = force_constants_from_file(args.filename)
+    print("Force constants data was loaded. Setting up dimensions...")
+
+    q_min = _get_q_distance(args.length_unit, args.q_min)
+    q_max = _get_q_distance(args.length_unit, args.q_max)
+    recip_length_unit = q_min.units
+
+    n_q_bins = ceil((args.q_max - args.q_min) / args.q_spacing)
+    q_bin_edges = np.linspace(q_min.magnitude, q_max.magnitude, n_q_bins + 1,
+                              endpoint=True) * recip_length_unit
+    q_bin_centers = (q_bin_edges[:-1] + q_bin_edges[1:]) / 2
+
+    # Use X-point modes to estimate frequency range, set up energy bins
+    # (Not Gamma in case there are only 3 branches; value would be zero!)
+
+    energy_bins, energy_unit = _get_energy_bins_and_units(
+        args.energy_unit,
+        fc.calculate_qpoint_phonon_modes(np.array([[0., 0., 0.5]]),
+                                         **calc_modes_kwargs),
+        args.ebins, emin=args.e_min, emax=args.e_max,
+        headroom=1.2)  # Generous headroom as we only checked one q-point
+
+    if args.weights in ('coherent',):
+        # Compute Debye-Waller factor once for re-use at each mod(q)
+        # (If temperature is not set, this will be None.)
+        if args.temperature is not None:
+            temperature = args.temperature * ureg['K']
+            dw = _get_debye_waller(temperature, fc, grid=args.grid,
+                                   grid_spacing=(args.grid_spacing
+                                                 * recip_length_unit),
+                                   **calc_modes_kwargs)
+        else:
+            temperature = None
+            dw = None
+
+    print(f"Sampling {n_q_bins} |q| shells between {q_min:~P} and {q_max:~P}")
+
+    z_data = np.empty((n_q_bins, len(energy_bins) - 1))
+
+    for q_index in tqdm(range(n_q_bins)):
+        q = q_bin_centers[q_index]
+
+        if args.npts_density is not None:
+            npts = ceil(args.npts_density * (q / recip_length_unit)**2)
+            npts = max(args.npts_min,
+                       min(args.npts_max, npts))
+        else:
+            npts = args.npts
+
+        if args.weights == 'dos':
+            spectrum_1d = sample_sphere_dos(
+                fc, q,
+                npts=npts, sampling=args.sampling, jitter=args.jitter,
+                energy_bins=energy_bins,
+                **calc_modes_kwargs)
```

Comment:
Not sure if this belongs in this PR (this probably should've gone in the `QpointFrequencies` PR), but in `sample_sphere_dos` you can probably use `calculate_qpoint_frequencies` to save some memory... Similarly for in `dos.py`, and in `intensity_map.py` if `-w dos`...

Reply:
Could have been another PR but did it here anyway